File: python/tests/test_fixtures.py
"""The cross-language contract: the same input reduced the same way gives the same numbers.

A digest mismatch here is a bug in whichever implementation moved, never a fixture to regenerate.
Fixtures are regenerated on the R side, deliberately, in their own commit, and only when
``inst/spec/representation.md`` changed with them.
"""

from __future__ import annotations

import csv
from pathlib import Path

import numpy as np
import pytest

from timegrain import digest_array, window_matrix

FIXTURES = Path(__file__).resolve().parents[2] / "inst" / "spec" / "fixtures"

SERIES_FILE = {"aligned": "series.csv", "offset": "series_offset.csv"}


def read_series(name):
    with (FIXTURES / SERIES_FILE[name]).open(newline="") as fh:
        rows = list(csv.DictReader(fh))
    return {"id": [r["id"] for r in rows],
            "time": [r["time"].replace("Z", "") for r in rows],
            "value": [float(r["value"]) for r in rows]}


def read_digests():
    with (FIXTURES / "digests.csv").open(newline="") as fh:
        return list(csv.DictReader(fh))


def read_edges(name):
    with (FIXTURES / "seasons.csv").open(newline="") as fh:
        rows = [r["edge"].replace("Z", "") for r in csv.DictReader(fh) if r["series"] == name]
    return np.asarray(rows, dtype="datetime64[s]")


def binning(name, window):
    if window != "astronomical":
        return window
    edges = read_edges(name)

    def astronomical(when):
        return edges[np.searchsorted(edges, when, side="right") - 1]

    return astronomical


@pytest.fixture(scope="module")
def series():
    return {name: read_series(name) for name in SERIES_FILE}


@pytest.mark.parametrize(
    "row", read_digests(),
    ids=lambda r: f"{r['series']}-{r['window']}-{r['year_start']}-{r['partial']}-{r['stat']}")
def test_digest_matches_the_r_side(series, row):
    x = window_matrix(series[row["series"]], "id", "time", "value",
                      window=binning(row["series"], row["window"]),
                      stats=row["stat"].split("+"), year_start=row["year_start"],
                      partial=row["partial"])
    # The shape is asserted before the digest, so a binning that puts the record into a different
    # number of bins is reported as that rather than as an unexplained hash mismatch.
    assert x.values.shape[0] == int(row["n_unit"])
    assert x.values.shape[1] == int(row["n_bin"])
    assert x.bins[0] == row["first_bin"]
    assert x.bins[-1] == row["last_bin"]
    assert int(x.bin_partial.sum()) == int(row["n_partial"])
    assert digest_array(x) == row["digest"]


def test_the_fixtures_cover_a_record_that_starts_on_no_bin_boundary():
    # A record beginning at midnight on the year_start anniversary puts every window in phase with
    # it, which is the one input on which a rule that keeps a partial leading bin and a rule that
    # never makes one agree. The contract is only a contract if it also carries the other case.
    rows = read_digests()
    assert {r["series"] for r in rows} == {"aligned", "offset"}
    offset = [r for r in rows if r["series"] == "offset"]
    assert {r["window"] for r in offset} == {"hour", "halfday", "day", "week", "month", "season",
                                             "year", "astronomical"}
    assert sum(int(r["n_partial"]) for r in offset) > 0
    assert {r["partial"] for r in rows} == {"keep", "drop"}
    assert len({r["year_start"] for r in rows}) > 1


def test_dropping_every_bin_is_an_error_rather_than_an_empty_representation(series):
    with pytest.raises(ValueError, match="no whole year"):
        window_matrix(series["offset"], "id", "time", "value", window="year", partial="drop")
    with pytest.raises(ValueError, match="must be"):
        window_matrix(series["offset"], "id", "time", "value", window="day", partial="sometimes")


def test_the_digest_is_the_lf_terminated_twelve_place_form_and_nothing_else():
    import hashlib
    values = np.array([1.0, -0.5]).reshape(2, 1, 1)
    expected = hashlib.md5(b"1.000000000000\n-0.500000000000\n").hexdigest()
    assert digest_array(values) == expected


def test_the_traversal_is_unit_fastest_then_bin_then_channel():
    values = np.arange(2 * 3 * 2, dtype=float).reshape(2, 3, 2)
    flat = values.flatten(order="F")
    assert list(flat[:2]) == [values[0, 0, 0], values[1, 0, 0]]
    assert flat[2] == values[0, 1, 0]

File: tests/testthat/test-fixtures.R
fixture_dir <- function() {
  installed <- system.file("spec", "fixtures", package = "timegrain")
  candidates <- c(installed, "../../inst/spec/fixtures", "inst/spec/fixtures")
  for (p in candidates) {
    if (nzchar(p) && file.exists(file.path(p, "digests.csv"))) {
      return(p)
    }
  }
  NULL
}

fixture_series <- function(dir, name) {
  file <- if (name == "aligned") "series.csv" else "series_offset.csv"
  s <- read.csv(file.path(dir, file), stringsAsFactors = FALSE)
  s$time <- as.POSIXct(s$time, format = "%Y-%m-%dT%H:%M:%SZ", tz = "UTC")
  s
}

fixture_binning <- function(dir, name, window) {
  if (window != "astronomical") {
    return(window)
  }
  edges <- read.csv(file.path(dir, "seasons.csv"), stringsAsFactors = FALSE)
  edges <- as.POSIXct(edges$edge[edges$series == name], format = "%Y-%m-%dT%H:%M:%SZ", tz = "UTC")
  function(when) edges[findInterval(as.numeric(when), as.numeric(edges))]
}

test_that("every representation matches the digest the Python side reads", {
  dir <- fixture_dir()
  skip_if(is.null(dir), "fixtures are not in the built package")

  expected <- read.csv(file.path(dir, "digests.csv"), stringsAsFactors = FALSE)
  series <- lapply(stats::setNames(nm = unique(expected$series)), fixture_series, dir = dir)

  for (i in seq_len(nrow(expected))) {
    row <- expected[i, ]
    label <- paste(row$series, row$window, row$year_start, row$partial, row$stat)
    x <- window_matrix(series[[row$series]], id, time, value,
                       window = fixture_binning(dir, row$series, row$window),
                       stats = strsplit(row$stat, "+", fixed = TRUE)[[1L]],
                       year_start = row$year_start, partial = row$partial)
    start <- attr(x, "bin_start")
    # The shape is asserted before the digest, so a binning that puts the record into a different
    # number of bins is reported as that rather than as an unexplained hash mismatch.
    expect_equal(dim(x)[1], row$n_unit, info = label)
    expect_equal(dim(x)[2], row$n_bin, info = label)
    expect_equal(format(start[1], "%Y-%m-%dT%H:%M:%SZ", tz = "UTC"), row$first_bin, info = label)
    expect_equal(format(start[length(start)], "%Y-%m-%dT%H:%M:%SZ", tz = "UTC"),
                 row$last_bin, info = label)
    expect_equal(sum(attr(x, "bin_partial")), row$n_partial, info = label)
    expect_identical(timegrain:::.digest_array(x), row$digest, info = label)
  }
})

test_that("the fixtures cover a record that starts on no bin boundary", {
  dir <- fixture_dir()
  skip_if(is.null(dir), "fixtures are not in the built package")
  expected <- read.csv(file.path(dir, "digests.csv"), stringsAsFactors = FALSE)

  # A record beginning at midnight on the year_start anniversary puts every window in phase with
  # it, which is the one input on which a rule that keeps a partial leading bin and a rule that
  # never makes one agree. The contract is only a contract if it also carries the other case.
  expect_true(all(c("aligned", "offset") %in% expected$series))
  offset <- expected[expected$series == "offset", ]
  expect_true(all(c("hour", "halfday", "day", "week", "month", "season", "year", "astronomical")
                  %in% offset$window))
  expect_gt(sum(offset$n_partial), 0)
  expect_true(all(c("keep", "drop") %in% expected$partial))
  expect_gt(length(unique(expected$year_start)), 1L)

  # Every window whose bin count the offset record splits differently from the aligned one is
  # pinned by a row of its own, so a change to either binning rule moves a digest here.
  aligned <- expected[expected$series == "aligned" & expected$stat == "mean" &
                        expected$partial == "keep" & expected$year_start == "09-01", ]
  expect_setequal(aligned$window,
                  c("hour", "halfday", "day", "week", "month", "season", "year", "astronomical"))
})

test_that("dropping every bin is an error rather than an empty representation", {
  t <- seq(as.POSIXct("2021-10-17 05:00:00", tz = "UTC"), by = "hour", length.out = 24 * 40)
  d <- data.frame(plot = "a", t = t, temp = sin(seq_along(t) / 24), stringsAsFactors = FALSE)
  expect_error(window_matrix(d, plot, t, temp, window = "year", partial = "drop"),
               "no whole year")
  expect_error(window_matrix(d, plot, t, temp, window = "day", partial = "sometimes"),
               "'arg' should be one of")
})

test_that("the digest is the LF-terminated twelve-place form and nothing else", {
  x <- array(c(1, -0.5), dim = c(2L, 1L, 1L))
  body <- charToRaw("1.000000000000\n-0.500000000000\n")
  f <- tempfile()
  con <- file(f, open = "wb")
  writeBin(body, con)
  close(con)
  expect_identical(timegrain:::.digest_array(x), unname(tools::md5sum(f)))
  unlink(f)
})
